### ISSUE TITLE
Early styling suggestion on fast failure to load v0.js

### ISSUE THREAD

**Issue by Gregable** [ampproject/amphtml#22543]
**Background**
AMP pages use a clever CSS animation to hide rendering until the AMP runtime has loaded, in order to avoid the dreaded flash of unstyled content:

(Simplified by removing the vendor-prefix variants):
```html
<style amp-boilerplate>
animation:-amp-start 8s steps(1,end) 0s 1 normal both}
@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}
</style>
```

For browsers which don't execute Javascript, there is a fallback:
```html
<noscript>
<style amp-boilerplate>
body {animation:none}
</style>
```

If Javascript is enabled, but the AMP javascript is slow to load, the 8 second animation unblocks the rendering after 8 seconds. A reasonable compromise.

**Problem**
Sometimes Javascript is enabled, and the AMP javascript fails to load in a manner which is detectable much earlier than 8 seconds. For example, if DNS fails to resolve to `cdn.ampproject.org`, a publisher's CORS header prevents loading the cross-origin resource, or an overzealous ad blocker prevents the request.

In these situations, it would be nice to "fail fast" and unblock the visibility of the page faster than 8 seconds.

**Proposal**
We can take advantage of the `onerror` attribute in browsers that have support and we can use this feature to "fail fast". For example:

```html
<script async src="https://cdn.ampproject.org/v0.js"                       
  onerror="document.getElementsByTagName('body')[0].style.cssText += 
           'webkit-animation:none;' +
           '-moz-animation:none;' +
           '-ms-animation:none;' + 
           'animation:none'"></script> 
```

This seems to work, though there is probably a shorter / more elegant version. I could use help in finding that minimal incantation.

I propose that we add this onerror attribute to all v0.js script tags in the AMP Cache and anywhere the AMP transformers run, such as the AMP Packager. We should also allow this exact onerror string in the AMP validator, though not require it.

@ampproject/wg-runtime for thoughts and @cramforce who originally designed the CSS timeout animation.

**Comment by cramforce** [comment]
I like the suggestion, but unfortunately it isn't compatible with CSP. Maybe we should do it anyway.

I have a very practical solution: Let's consider reducing the timeout to something much shorter like 3s.

**Comment by Gregable** [comment]
> I like the suggestion, but unfortunately it isn't compatible with CSP.

Ah ok.

> Maybe we should do it anyway.

What do you mean? For browsers that don't support CSP or something?

>  Let's consider reducing the timeout to something much shorter like 3s.

Seems good to me.

**Comment by jridgewell** [comment]
3s seems fine to me.

**Comment by Gregable** [comment]
Will change to 3s.

We could still do this "fail fast" though by updating the CSP. Roughly:

1) Rather than onerror="", add new inline script to the bottom of every amp document that attaches the onerror to the v0.js script tag.
2) Add a `content-security-policy: ... script-src sha256-8oqAbZljk3byxMyw...` to the AMP Cache CSP that matches the new inline script.

The same inline script could be allowed by the AMP Validator and used in Signed Exchanges.

Downside is that we add extra header bytes to every response to carry the sha hash in the CSP.

**Comment by choumx** [comment]
With 3s, I reliably see FOUC on examples/article.amp.html with Chrome's "Slow 3G" preset. I think we should just reduce to 5s or so.

Nice CSP fix. You can just do `document.body` to shave a few chars. 

I also think we can ignore the `-moz-` and `-ms-` vendor prefixes based on [market share](https://caniuse.com/#feat=css-animation). Firefox 5-15 uses `-moz-` and is 0.05%. IE10 appears to be the first IE that supports CSS animations and uses non-prefixed (not sure `-ms-animation` was ever used).

**Comment by cramforce** [comment]
With adding the CSP fix (was this tested empirically with a content blocker
browser extension?) I think going to 5s is probably fine since we are
eliminating a major source of timeouts.

Will: Do you think no FOUC is worth the extra 2s load time for a user on a
slow connection?

**Comment by Gregable** [comment]
Hm, I just pushed out the 3s fix to the AMP validator code (it went in to the production validator last week), before seeing this comment. If you want, I can roll it back, but should do so ASAP.

**Comment by Gregable** [comment]
Discussed with @choumx offline, will roll back until a consensus has been reached.

**Comment by choumx** [comment]
> Will: Do you think no FOUC is worth the extra 2s load time for a user on a
slow connection?

Yea, I think that such users have different latency expectations. 

We should continue dropping this over time though. [A quick search shows](https://economictimes.indiatimes.com/industry/telecom/telecom-news/user-base-for-2g-3g-shrinking-as-upwardly-mobile-take-to-4g/articleshow/68072430.cms) over half of India on 2G in 2018, but 4G predicted to take over in two years. :)

**Comment by Gregable** [comment]
This thread stalled a bit on determining the best timeout value for the CSS animation. I'd like to revisit the `onerror` handler. I recommend that if folks are still interested in the timeout we start a new issue for that discussion.



To add some more detail, I did some testing with the following setup. I added to a document head the following:

```
<script>document.querySelector("script[src*='v0.js']").onerror=function(){document.getElementsByTagName('body')[0].style.cssText += 'webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none';}</script>
```

This doesn't execute due to CSP as mentioned above, but also has no negative effects. I think added to the CSP header's `script-src` the stanza `'sha256-qzw+iiMAXBAsFo7HI5xT6iXbhaPwLtOcRIFHEiWURHw='` which matches the script tag contents above.

This runs the above script tag and successfully unhides the document when v0.js fails to load.



For reference, the pretty-print version of the script is:

```
<script>
document.querySelector("script[src*='v0.js']").onerror = function() {
  document.getElementsByTagName('body')[0].style.cssText +=
    'webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none';
}
</script>
```

The suggested change is simply to allow (but not require) this script tag in validation, and add it to all AMP Cache pages, along with the matching CSP SHA header. The existing CSS animation would remain in place, and remain required. Documentation of the boilerplate would be updated to include the script tag.



It has already been suggested by @dvoytenko that we could shorten with:

```
document.body.style.cssText +=
```

If anyone has suggestions on ways to shorten further before we select the final string, feel free to add them here.

Also if you see any drawbacks or security considerations, let me know. @ampproject/wg-security-privacy

**Comment by jridgewell** [comment]
Sounds fine to me.

Nit: This is likely the shortest solution, and requires no changes if we update the boilerplate CSS with something else later on.

```html
  <script async src="https://cdn.ampproject.org/v0.js" onerror="document.querySelector('style.amp-boilerplate').textContent=''"></script>
```

**Comment by Gregable** [comment]
@jridgewell as far as I could determine, the CSP sha256 stanza doesn't work with strings inside an `onerror` attribute. It's possible I made some other mistake, but it didn't work with my quick tests.

**Comment by dvoytenko** [comment]
@Gregable @jridgewell even if we can't shad256 an inline `onerror` handler, we could combine these two ideas:

```
<style amp-boilerplate>....</style>
...
<script>
document.querySelector("script[src*='v0.js']").onerror = function() {
  document.querySelector('style[amp-boilerplate]').textContent='';
}
</script>
```

It kind of has a logic feel to it and it looks similar to script/noscript. Though in reality the difference is probably pretty small.

**Comment by jridgewell** [comment]
I was using `'unsafe-hashes'`, but it seems Safari doesn't support it: https://humdrum-olivine.glitch.me/

Let's do @dvoytenko's https://github.com/ampproject/amphtml/issues/22543#issuecomment-571370680.

**Comment by cramforce** [comment]
We will need to enforce that the onerror handler is added immediately after the script element in the source. Otherwise this would be subject to race conditions where the parser halts between loading the script and parsing the doc, and then fails to load the script before the onerror handler can be added.

**Comment by Gregable** [comment]
@cramforce My understanding is that the worst that happens if out of order is that the onerror doesn't happen, so we're no worse than today.

I'm tempted not to validate the ordering out of simplicity. The validator doesn't have support for rules regarding sibling ordering, so that would complicate implementation a tiny bit. I imagine most implementations will pick this up through a cache transformation or the optimizer, which we can implement to order correctly out of the box.

**Comment by dvoytenko** [comment]
Hmm. I think the ordering check is relatively unavoidable here. This script has to be after both `amp-boilerplate` and `v0.js` if use do `style[amp-boilerplate].textContent = ''` and just `v0.js` if we use `body.style.X = ''`. That's partially why `onerror` would be very attractive but since we can't do hash in that case...

**Comment by Gregable** [comment]
Talked with @dvoytenko offline and understand the problem better. 

1) If the inline script executes before the `v0.js` script tag is parsed, attaching onerror will fail. This can be avoided by placing the inline script after `v0.js`.

1) If the onerror event occurs before the inline script is executed, the onerror handler won't run. This can be avoided by placing the inline script before `v0.js` but that will make attaching fail, see (1).

1) If the onerror handler occurs before the body tag is parsed, the handler won't do anything. This can be avoided by placing the inline script after the `body` tag, but that makes (2) more likely to fail.

These events are very race sensitive. Most likely placing the inline script immediately after `v0.js` or immediately after `body` would increase the hit rate the most, but no position is guaranteed.

I'm going to recheck my work on adding a SHA hash that matches the `onerror` attribute first and see if I can get that working, as it solves (1) and (2) for sure.

**Comment by twifkak** [comment]
Could this be solved with Mutation Observer to wait for the script and body
nodes to be created?
https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver

**Comment by Gregable** [comment]
`onerror` does work in Chrome with the `unsafe-hashes` that @jridgewell pointed out and I had initially missed. Also works in Firefox. I'll trust @jridgewell that it doesn't work in Safari as I don't have a device to test.

**Comment by cramforce** [comment]
@twifkak MutationObserver is a bit overkill.

I think moving the script element right after v0 and then adding a small `DOMContentLoaded` listener to execute the body change will be fine.

**Comment by Gregable** [comment]
If we use the `textContent` solution Justin proposed, we don't care when the `<body>` loads. The minimal solution then is:

```
<script amp-onerror>document.querySelector("script[src*='v0.js']").onerror=function(){document.querySelector('style[amp-boilerplate]').textContent=''}</script>
```

Pretty print:

```
<script amp-onerror>
document.querySelector("script[src*='v0.js']").onerror = function() {
  document.querySelector('style[amp-boilerplate]').textContent = ''
}
</script>
```

And ensure all of:

1. `sha256-v5GILL6vYnmgSPupWB8RrkBejUN8fsnyUZVQAjXGwx8=` is in the CSP header.
1. `<style amp-boilerplate>` comes before `<script src=v0.js>`.
1. `<script src=v0.js>` comes **immediately** before `<script amp-onerror>`.

We might be able to shave a couple more bytes off that, but it doesn't seem necessary. I added the `amp-onerror` attribute to make it simple to produce good validation errors.

The only case where this fails is when the `onerror` triggers before the `<script amp-onerror>` is executed, which should be rare if we ensure (3).

**Comment by jridgewell** [comment]
SGTM.

Should we also allow Pubs to include this script in their pages? Or only as a cache transformation?

**Comment by Gregable** [comment]
Good Question: The validator has two modes (transformed vs. not). Transformed validation handles documents in cache, signed exchanges, and amp optimizer output. The AMP transform code paths already output `<style amp-custom>` before `<script v0.js>`.

Forcing the ordering is fairly easy in the transform code paths, so we'll add them to the AMP Cache, Optimizer, and Packager. However, validator implementation requires a bit of additional work.

I propose the easiest to implement Option A:

- Validator allows `<script amp-onerror>` in **both modes**, optionally.
- Validator **doesn't validate ordering**.

Option B:
- Validator allows `<script amp-onerror>` **only in transformed mode**, optionally.
- Validator doesn't validate ordering.

Option C:
- Validator allows `<script amp-onerror>` in both modes, optionally.
- Validator **validates ordering** only if `<script amp-onerror>` is present.


I don't feel the validator validating ordering is important:
 - If out of order, worst case is that we have the same as today.
 - For transformed mode, there's no reason we'd ever be out of order since this is generated code, so the extra work to validate ordering seems unnecessary.
 - For non-transformed mode, it doesn't make sense to remove a document from an AMP Cache.
 - For non-transformed mode, maybe a publisher would like to know they made a mistake, but this is not typically the role of the validator. The document isn't really broken, seems like a bad reason to remove from caches.
 - For non-transformed mode, we'd need to implement a oddly specific one-off rule: `<style amp-custom>` must come before `<script v0.js>`, but only if `<script amp-onerrror>` is present.

**Comment by jridgewell** [comment]
I'd be fine with Option A.

**Comment by dvoytenko** [comment]
I think option c might be a bit better, but option a is ok as well.

**Comment by molnarg** [comment]
You could also use `document.currentScript.previousElementSibling` to save 2 more bytes :) `with(document)` might save a couple more in exchange of using a deprecated feature.

Also IIRC inline script tag execution (no matter where it is on the page) comes before any JS event can be fired, because it's done as part of the initial parsing of the page. There's no network latency in this example and still executes correctly: data:text/html,<script src=x></script> <script>document.currentScript.previousElementSibling.onerror=function(){alert(1)}</script>

**Comment by Gregable** [comment]
As an update, this was implemented and rolled out in the Google AMP Cache. The final version is:

```
<script amp-onerror>document.querySelector("script[src*='/v0.js']").onerror=function(){document.querySelector('style[amp-boilerplate]').textContent=''}</script>
```

It was combined with a matching CSP change:

```
sha256-5CxqAdDXlHviOy7zxeRpMobzRK/JNpLvkS+k8Zj3L3A=
```

The remaining work is to allow the same in the validator and perhaps add this to documentation. I think this reduces the remaining priority, so I'll update the labels accordingly.

**Comment by twifkak** [comment]
After allowing in the validator, it would be good to add support to https://github.com/ampproject/amppackager and https://github.com/ampproject/amp-toolbox/tree/master/packages/optimizer.

**Comment by honeybadgerdontcare** [comment]
Filed https://github.com/ampproject/amp.dev/issues/5442 for documentation as the validator will soon allow:

```
<script amp-onerror>document.querySelector("script[src*='/v0.js']").onerror=function(){document.querySelector('style[amp-boilerplate]').textContent=''}</script>
```

**Comment by honeybadgerdontcare** [comment]
Filed also for AMP Packager and AMP Optimizer.

Remaining work for this issue is adding to the Validator which is currently in review.

**Comment by sebastianbenz** [comment]
@westonruter @schlessera FYI for the Wordpress Optimizer

**Comment by honeybadgerdontcare** [comment]
Fixed in #32989

**Comment by sebastianbenz** [comment]
Do we need to handle the module case here as well? `script[src*='/v0.js'],script[src*='/v0.mjs]`?

**Comment by schlessera** [comment]
> Do we need to handle the module case here as well? script[src*='/v0.js'],script[src*='/v0.mjs]?

Given that the `.mjs` file is the new default, I'd say yes, this needs to be taken into account. However, it looks like the validator has the entire `amp-onerror` content hard-coded, so it will need another change to allow for that.

**Comment by schlessera** [comment]
I'm wondering whether the validator should also allow for other version string than `v0` here...?

**Comment by westonruter** [comment]
> Do we need to handle the module case here as well? `script[src*='/v0.js'],script[src*='/v0.mjs]`?

Yes, as for AMP pages optimized with ESM scripts, the current `script[amp-onerror]` will do absolutely nothing for modern browsers. Only IE11 will get the error handling since only it will attempt to load `v0.js`. Other browsers will attempt to load `v0.mjs`.

It's not just a matter of adding a selector, though. We'd need to create a loop. So instead of:

```js
document.querySelector("script[src*='/v0.js']").onerror=function(){document.querySelector('style[amp-boilerplate]').textContent=''}
```

We'd need to do:

```js
Array.prototype.forEach.call(
	document.querySelectorAll( "script[src*='/v0.js'],script[src*='/v0.mjs']" ),
	function ( script ) {
		script.onerror = function () {
			document.querySelector( 'style[amp-boilerplate]' ).textContent = '';
		};
	}
);
```

The use of `Array#forEach` is required over `NodeList#forEach` since the former is [supported in IE≥9](https://caniuse.com/mdn-javascript_builtins_map_foreach) but the latter is [not supported in IE11](https://caniuse.com/mdn-api_nodelist_foreach). 

Note that `Array#forEach` is supported in IE≥9, but `NodeList#forEach` . Minified it could be:

```js
Array.prototype.forEach.call(document.querySelectorAll("script[src*='/v0.js'],script[src*='/v0.mjs']"),function(s){s.onerror=function(){document.querySelector('style[amp-boilerplate]').textContent=''}})
```

**Comment by westonruter** [comment]
I'm not sure how this would be achieved in the validator while maintaining validity for the current `v0.js` script:

https://github.com/ampproject/amphtml/blob/735b7cbc40fbe702dff2821d44f479b7a389e0c4/validator/validator-main.protoascii#L3370-L3387

Instead of `mandatory_cdata` it seems we could use an `allowed_cdata_regex` constraint use and `|` the patterns together like is common for `value_regex`, but doesn't seem to only exist. Only `disallowed_cdata_regex` is a thing.

I suppose the entire tag spec could be copied to live alongside the current one:

```
tags: {
  html_format: AMP
  tag_name: "SCRIPT"
  spec_name: "script amp-onerror (including ESM)"
  descriptive_name: "script amp-onerror (including ESM)"
  unique: true
  mandatory_parent: "HEAD"
  attrs: {
    name: "amp-onerror"
    value: ""
    mandatory: true
    dispatch_key: NAME_VALUE_DISPATCH
  }
  cdata: {
    mandatory_cdata: "Array.prototype.forEach.call(document.querySelectorAll(\"script[src*='/v0.js'],script[src*='/v0.mjs']\"),function(s){s.onerror=function(){document.querySelector('style[amp-boilerplate]').textContent=''}})"
  }
}
```

**Comment by jridgewell** [comment]
Yes, this is related to b/163852894. The official (Google internal) module/nomodule variant is:

```html
<script amp-onerror>[].slice.call(document.querySelectorAll("script[src*='/v0.js'],script[src*='/v0.mjs']")).forEach(function(s){s.onerror=function(){document.querySelector('style[amp-boilerplate]').textContent=''}})</script>
```

@honeybadgerdontcare: How would you suggest we add this to validator?

**Comment by westonruter** [comment]
PR opened to update PHP Optimizer: https://github.com/ampproject/amp-toolbox-php/pull/211.

**Comment by westonruter** [comment]
Reopening per https://github.com/ampproject/amphtml/pull/34531#issuecomment-849733408:

> There seems to be a problem with the validator change, namely that the old v0.js version being present in a transformed AMP page now causes a validation error: https://testsite.cmsdevrel.com/web-stories/thailand-copy/
>
> Transformed AMP should allow either the new version or the old version, while non-transformed AMP only needs to allow the old version.

**Comment by honeybadgerdontcare** [comment]
The issue here is the transformed AMP document does not have the updated amp-onerror that handles both v0.js and v0.mjs. It's also not adding module/nomodule support. I'll take a look later today to support the old amp-onerror for transformed AMP. The work around is for the publisher to use the new amp-onerror script content.

**Comment by westonruter** [comment]
The story page is transformed AMP, but it doesn't include the ESM scripts (context: https://github.com/google/web-stories-wp/issues/7649). The Optimizers allow individual transformers to be disabled, and we've seen cases where site owners need to turn off ESM due to incompatibilities with caching layers that aren't yet aware of `type="module"` on `script` tags. So really the new `script[amp-onerror]` should not be required based on whether the page is transformed, but rather if it has ESM scripts (modue/nomodule support).

**Comment by westonruter** [comment]
But both should be valid at least for now since we haven't pushed out the Optimizer updates to add the new version yet.

**Comment by trby** [comment]
I'm currently dealing with this error that popped up in my search console 3 days ago: **The mandatory text inside tag ‘script amp-onerror’ is missing or incorrect.** It’s affecting all my amp pages. I'm not sure how to fix this problem. Any help is appreciated.

**Comment by westonruter** [comment]
@trby Hi there. I see you opened a support topic for the WordPress AMP plugin as well: https://wordpress.org/support/topic/the-mandatory-text-inside-tag-script-amp-onerror-is-missing-or-incorrect-4/

Other users have also reported this problem:

* https://wordpress.org/support/topic/script-amp-onerror-2/
* https://wordpress.org/support/topic/the-mandatory-text-inside-tag-script-amp-onerror-is-missing-or-incorrect-3/
* https://wordpress.org/support/topic/script-amp-onerror-is-missing-or-specified-incorrectly/

The underlying problem is with the AMP Validator which should be fixed this week, hopefully tomorrow. In the meantime, you can just disable the `amp-onerror` functionality by activating this mini plugin: https://gist.github.com/westonruter/d13eb0175a22f83c41531e7d1a7071fd

**Comment by trby** [comment]
Thanks for your quick reply. I installed the mini plugin but when I try to revalidate an AMP page, the error still persists. I'll wait for the permanent fix. Again thanks for your help.

**Comment by schlessera** [comment]
I think it might be good to have an attribute on the `<html>` tag set to cheaply detect whether ESM is used or not (without regexing the content or iterating over the nodes).

This should probably be `i-amphtml-module`/`i-amphtml-esm` to denote ESM support, instead of denoting no ESM support, as the modules are only added by the optimizer, and hand-coding AMP (which would have no ESM) should not need the additional tag to be valid.

**Comment by schlessera** [comment]
Adding the above extra attribute would also mean the validation rules can easily distinguish between ESM/no-ESM, just as they do with transformed/not-transformed right now.

**Comment by honeybadgerdontcare** [comment]
There are two components to validation: the JavaScript validator released publicly and internal versions that AMP caches use. The Google AMP Cache has added the validation change for Transformed AMP today and will be in production within the week. The JavaScript validator has added the same change in #34634. The JavaScript validator served on cdn.ampproject.org will be updated today with this change.

This means DevTools like validator.amp.dev will have the change by tomorrow and the Google AMP Cache (where Search Console gets it's reports) by the end of the week.

The immediate work around is to either remove the amp-onerror module as @westonruter mentioned or update that module to use the cdata for Transformed AMP.

**Comment by trby** [comment]
> @trby Hi there. I see you opened a support topic for the WordPress AMP plugin as well: https://wordpress.org/support/topic/the-mandatory-text-inside-tag-script-amp-onerror-is-missing-or-incorrect-4/
> 
> Other users have also reported this problem:
> 
> * https://wordpress.org/support/topic/script-amp-onerror-2/
> * https://wordpress.org/support/topic/the-mandatory-text-inside-tag-script-amp-onerror-is-missing-or-incorrect-3/
> * https://wordpress.org/support/topic/script-amp-onerror-is-missing-or-specified-incorrectly/
> * https://wordpress.org/support/topic/amp-optimization-could-not-be-completed-due-to-amp-experiment/#post-14504642
> 
> The underlying problem is with the AMP Validator which should be fixed this week, hopefully tomorrow. In the meantime, you can just disable the `amp-onerror` functionality by activating this mini plugin: https://gist.github.com/westonruter/d13eb0175a22f83c41531e7d1a7071fd

Hi there! Does the 2.1.3 version fix this problem? Is it safe to uninstall the mini plugin (Disable amp-onerror Handler)? Thanks for your help.

**Comment by westonruter** [comment]
@trby The fix was made to the validator, so the mini plugin is no longer needed.

**Comment by westonruter** [comment]
Nevertheless, AMP plugin v2.1.3 also adds support for amp-onerror on transformed pages (with ESM scripts). So be sure to update to that version as well.
